File: xiaolearn_app/scripts/fix-dict-pinyins.mjs
/**
 * fix-dict-pinyins.mjs
 * ---------------------
 * Régénère les pinyins défaillants dans les chunks du dictionnaire CFDICT.
 *
 * Contexte : audit a montré
 *   - Entrées principales : ~4 % de divergences vs pinyin canonique
 *   - Exemples (phrases) : ~99 % de divergences (transcription erronée
 *     par l'outil source à l'origine du dataset).
 *
 * Ce script utilise `pinyin-pro` (déjà en dépendance) pour générer le
 * pinyin canonique avec tons à partir des hanzi. Trois modes :
 *
 *   --target=examples (défaut) : ne touche QUE aux exemples (gain maximal,
 *                                risque minimal de casser des entrées dont
 *                                la lecture stockée est correcte mais rare).
 *   --target=entries           : régénère le pinyin principal de chaque
 *                                entrée (à utiliser avec --confirm-entries
 *                                car risque de perdre des lectures rares).
 *   --target=all               : les deux.
 *   --dry-run                  : montre ce qui serait changé sans écrire.
 *   --report=path.json         : exporte un rapport JSON des modifications.
 *
 * Usage :
 *   node scripts/fix-dict-pinyins.mjs --dry-run
 *   node scripts/fix-dict-pinyins.mjs --target=examples
 *   node scripts/fix-dict-pinyins.mjs --target=all --confirm-entries
 *
 * Sortie : modifie les chunks `public/data/hors-hsk/chunk-*.json` (et
 * éventuellement HSK) en place. Garde les autres champs (translationFr,
 * audio, examples translation, etc.) intacts.
 */

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { pinyin } from 'pinyin-pro';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const PROJECT_ROOT = path.resolve(__dirname, '..');

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------
const argv = process.argv.slice(2);
const has = (name) => argv.includes(`--${name}`);
const getVal = (name) =>
  argv.find((a) => a.startsWith(`--${name}=`))?.slice(`--${name}=`.length);

const dryRun = has('dry-run');
const target = getVal('target') || 'examples';
const confirmEntries = has('confirm-entries');
const reportPath = getVal('report');

if (target !== 'examples' && target !== 'entries' && target !== 'all') {
  console.error(`❌ --target invalide : ${target}. Valeurs : examples | entries | all`);
  process.exit(1);
}
if ((target === 'entries' || target === 'all') && !confirmEntries && !dryRun) {
  console.error(
    `❌ Modifier les pinyins d'entrées (--target=${target}) écrase potentiellement\n` +
      `   des lectures rares valides. Ajoute --confirm-entries pour confirmer,\n` +
      `   ou utilise --dry-run pour prévisualiser sans toucher au disque.`
  );
  process.exit(2);
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------
const CJK_RE = /[一-鿿]/;

/** Pinyin canonique avec tons (ex: "我爱你" → "wǒ ài nǐ"). */
function canonicalPinyin(hanzi) {
  if (!hanzi || !CJK_RE.test(hanzi)) return '';
  // pinyin-pro retourne un string séparé par espaces. Pour les chars non-CJK
  // (ponctuation, latin), on les ignore en activant nonZh: 'consecutive'.
  return pinyin(hanzi, {
    toneType: 'symbol',
    type: 'string',
    nonZh: 'consecutive',
    v: false
  })
    .replace(/\s+/g, ' ')
    .trim();
}

/** Normalise pour comparaison : sans diacritiques, sans espaces, casse. */
function normCompare(s) {
  return s
    .normalize('NFD')
    .replace(/[̀-ͯ]/g, '')
    .replace(/\s+/g, '')
    .toLowerCase();
}

// ---------------------------------------------------------------------------
// Process chunks
// ---------------------------------------------------------------------------
function processChunk(filePath, isHorsHsk) {
  const data = JSON.parse(fs.readFileSync(filePath, 'utf8'));
  const entries = isHorsHsk ? data : Object.values(data);
  let entriesFixed = 0;
  let examplesFixed = 0;
  const samples = [];

  for (const e of entries) {
    if (!e || typeof e !== 'object') continue;

    // 1. Pinyin de l'entrée principale
    if ((target === 'entries' || target === 'all') && e.hanzi && CJK_RE.test(e.hanzi)) {
      const canon = canonicalPinyin(e.hanzi);
      if (canon && normCompare(canon) !== normCompare(e.pinyin || '')) {
        if (samples.length < 5) {
          samples.push({ kind: 'entry', hanzi: e.hanzi, old: e.pinyin, new: canon });
        }
        e.pinyin = canon;
        entriesFixed++;
      }
    }

    // 2. Pinyin des exemples
    if ((target === 'examples' || target === 'all') && Array.isArray(e.examples)) {
      for (const ex of e.examples) {
        if (!ex.chinese || !CJK_RE.test(ex.chinese)) continue;
        const canon = canonicalPinyin(ex.chinese);
        if (!canon) continue;
        if (normCompare(canon) === normCompare(ex.pinyin || '')) continue;
        if (samples.length < 5) {
          samples.push({
            kind: 'example',
            chinese: ex.chinese,
            old: ex.pinyin,
            new: canon
          });
        }
        ex.pinyin = canon;
        examplesFixed++;
      }
    }
  }

  return { entries, entriesFixed, examplesFixed, samples };
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------
function main() {
  console.log(`🔧 Fix pinyins dictionary${dryRun ? ' · DRY-RUN' : ''}`);
  console.log(`   Cible : ${target}`);
  console.log('');

  let totalEntries = 0;
  let totalExamples = 0;
  const reportEntries = [];
  const reportExamples = [];
  let firstSamples = [];

  // Hors-HSK : 110 chunks au format array
  const horsHskChunks = fs
    .readdirSync(path.join(PROJECT_ROOT, 'public', 'data', 'hors-hsk'))
    .filter((f) => f.startsWith('chunk-') && f.endsWith('.json'))
    .sort();

  console.log(`📖 Hors-HSK : ${horsHskChunks.length} chunks`);
  for (const c of horsHskChunks) {
    const fp = path.join(PROJECT_ROOT, 'public', 'data', 'hors-hsk', c);
    const result = processChunk(fp, true);
    if (!dryRun && (result.entriesFixed > 0 || result.examplesFixed > 0)) {
      fs.writeFileSync(fp, JSON.stringify(result.entries, null, 0));
    }
    totalEntries += result.entriesFixed;
    totalExamples += result.examplesFixed;
    if (firstSamples.length < 8) {
      firstSamples = firstSamples.concat(result.samples).slice(0, 8);
    }
    if (result.entriesFixed > 0 || result.examplesFixed > 0) {
      console.log(
        `   ${c}: ${result.entriesFixed} entrée(s), ${result.examplesFixed} exemple(s)`
      );
    }
  }

  // HSK : chunks au format map { id: entry }. On traite TOUJOURS les HSK
  // pour les exemples (gros volume ~22k), et SI demandé pour les entrées.
  {
    const hskLevels = ['hsk1', 'hsk2', 'hsk3', 'hsk4', 'hsk5', 'hsk6', 'hsk7'];
    for (const lvl of hskLevels) {
      const dir = path.join(PROJECT_ROOT, 'public', 'data', 'dictionary', 'hsk', lvl);
      if (!fs.existsSync(dir)) continue;
      const chunks = fs
        .readdirSync(dir)
        .filter((f) => f.startsWith('chunk-') && f.endsWith('.json'))
        .sort();
      console.log(`📚 ${lvl} : ${chunks.length} chunks`);
      for (const c of chunks) {
        const fp = path.join(dir, c);
        const raw = JSON.parse(fs.readFileSync(fp, 'utf8'));
        let entriesFixed = 0;
        let examplesFixed = 0;
        for (const e of Object.values(raw)) {
          // Pinyin entrée : seulement si target = entries|all
          if (
            (target === 'entries' || target === 'all') &&
            e.hanzi &&
            CJK_RE.test(e.hanzi)
          ) {
            const canon = canonicalPinyin(e.hanzi);
            if (canon && normCompare(canon) !== normCompare(e.pinyin || '')) {
              e.pinyin = canon;
              entriesFixed++;
            }
          }
          // Pinyin exemples : si target = examples|all
          if ((target === 'examples' || target === 'all') && Array.isArray(e.examples)) {
            for (const ex of e.examples) {
              if (!ex.chinese || !CJK_RE.test(ex.chinese)) continue;
              const canon = canonicalPinyin(ex.chinese);
              if (canon && normCompare(canon) !== normCompare(ex.pinyin || '')) {
                ex.pinyin = canon;
                examplesFixed++;
              }
            }
          }
        }
        if (!dryRun && (entriesFixed > 0 || examplesFixed > 0)) {
          fs.writeFileSync(fp, JSON.stringify(raw, null, 0));
        }
        totalEntries += entriesFixed;
        totalExamples += examplesFixed;
        if (entriesFixed > 0 || examplesFixed > 0) {
          console.log(`   ${c}: ${entriesFixed} entrée(s), ${examplesFixed} exemple(s)`);
        }
      }
    }
  }

  console.log('');
  console.log(`Résumé :`);
  console.log(`  Entrées corrigées : ${totalEntries}`);
  console.log(`  Exemples corrigés : ${totalExamples}`);
  console.log('');
  if (firstSamples.length > 0) {
    console.log(`Exemples de modifications :`);
    for (const s of firstSamples) {
      console.log(`  [${s.kind}] ${s.hanzi || s.chinese}`);
      console.log(`    avant : ${s.old}`);
      console.log(`    après : ${s.new}`);
    }
  }
}

main();

File: xiaolearn_app/scripts/fix-dict-pinyins.py
#!/usr/bin/env python3
"""
fix-dict-pinyins.py
--------------------
Version Python du script de correction des pinyins, avec segmentation par
MOTS via jieba (au lieu de char-by-char). Format de sortie cohérent avec
ce que les humains lisent : "使用" → "shǐyòng" (joint), pas "shǐ yòng".

Pipeline :
  1. jieba.cut() segmente "我使用电脑" → ["我", "使用", "电脑"]
  2. pypinyin convertit chaque segment en pinyin avec tons
  3. Les chars d'un même mot sont collés (shǐyòng), les mots séparés par
     espaces, ponctuation/digits passent à travers
  4. Sandhi tonale appliquée par pypinyin (yī → yí, bù → bú)

Usage :
  python3 scripts/fix-dict-pinyins.py --dry-run
  python3 scripts/fix-dict-pinyins.py --target=examples
  python3 scripts/fix-dict-pinyins.py --target=all --confirm-entries

Cibles :
  examples (défaut) : régénère uniquement les pinyins d'exemples
  entries           : régénère les pinyins d'entrées principales
                      (requiert --confirm-entries en mode réel)
  all               : les deux
"""

import json
import re
import sys
import argparse
import unicodedata
from pathlib import Path
import jieba
from pypinyin import pinyin, Style

PROJECT_ROOT = Path(__file__).resolve().parent.parent
HORS_HSK_DIR = PROJECT_ROOT / "public" / "data" / "hors-hsk"
HSK_DIR = PROJECT_ROOT / "public" / "data" / "dictionary" / "hsk"

CJK_RE = re.compile(r"[一-鿿]")

# Désactive les logs jieba
jieba.setLogLevel(60)


def canonical_pinyin(text: str) -> str:
    """
    Segmente le texte avec jieba puis génère le pinyin canonique :
    - Caractères CJK d'un même mot collés (使用 → shǐyòng)
    - Mots séparés par un espace
    - Ponctuation et non-CJK conservés tels quels
    - Sandhi tonale appliquée
    """
    if not text:
        return ""
    parts = []
    for seg in jieba.lcut(text):
        if CJK_RE.search(seg):
            # Récupère la forme tonale (avec sandhi) char-by-char, puis colle
            syllables = [p[0] for p in pinyin(seg, style=Style.TONE)]
            parts.append("".join(syllables))
        else:
            parts.append(seg)
    # Joindre avec espace, mais éviter doublons d'espaces (jieba peut sortir des espaces vides)
    out = " ".join(parts)
    out = re.sub(r"\s+", " ", out).strip()
    # Décoller la ponctuation collée à un espace : on veut "shǐyòng 。" pas "shǐ yòng。"
    # En fait l'inverse : on veut collée. Mais jieba les sort comme tokens séparés.
    # On rejoint les ponctuations qui suivent un mot pinyin sans espace.
    out = re.sub(r"\s+([，。！？、；：·…—""''（）《》])", r"\1", out)
    out = re.sub(r"([""''（《])\s+", r"\1", out)
    return out


def norm_compare(s: str) -> str:
    """Normalisation pour comparer : sans diacritiques, sans espaces, casse."""
    s = unicodedata.normalize("NFD", s)
    s = "".join(c for c in s if not unicodedata.combining(c))
    s = re.sub(r"\s+", "", s)
    return s.lower()


def process_entries(entries, target, samples):
    """Traite une liste d'entries et retourne (n_entries_fixed, n_examples_fixed)."""
    n_entries = 0
    n_examples = 0
    for e in entries:
        if not isinstance(e, dict):
            continue
        # Pinyin entrée principale
        hanzi = e.get("hanzi", "")
        if target in ("entries", "all") and hanzi and CJK_RE.search(hanzi):
            canon = canonical_pinyin(hanzi)
            if canon and norm_compare(canon) != norm_compare(e.get("pinyin", "")):
                if len(samples) < 8:
                    samples.append(
                        {
                            "kind": "entry",
                            "hanzi": hanzi,
                            "old": e.get("pinyin", ""),
                            "new": canon,
                        }
                    )
                e["pinyin"] = canon
                n_entries += 1
        # Pinyins examples
        if target in ("examples", "all"):
            for ex in e.get("examples", []) or []:
                ch = ex.get("chinese", "")
                if not ch or not CJK_RE.search(ch):
                    continue
                canon = canonical_pinyin(ch)
                if not canon:
                    continue
                if norm_compare(canon) == norm_compare(ex.get("pinyin", "")):
                    continue
                if len(samples) < 8:
                    samples.append(
                        {
                            "kind": "example",
                            "chinese": ch[:60],
                            "old": ex.get("pinyin", "")[:80],
                            "new": canon[:80],
                        }
                    )
                ex["pinyin"] = canon
                n_examples += 1
    return n_entries, n_examples


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--dry-run", action="store_true")
    parser.add_argument(
        "--target",
        choices=["entries", "examples", "all"],
        default="examples",
        help="Quel pinyin régénérer",
    )
    parser.add_argument(
        "--confirm-entries",
        action="store_true",
        help="Confirme l'écrasement des pinyins d'entrées (risque sur lectures rares)",
    )
    args = parser.parse_args()

    if args.target in ("entries", "all") and not args.confirm_entries and not args.dry_run:
        print(
            f"❌ --target={args.target} : ajoute --confirm-entries pour confirmer\n"
            f"   (les lectures rares peuvent être écrasées). Ou utilise --dry-run.",
            file=sys.stderr,
        )
        sys.exit(2)

    print(f"🔧 Fix pinyins (Python + jieba){'  · DRY-RUN' if args.dry_run else ''}")
    print(f"   Cible : {args.target}")
    print()

    total_entries = 0
    total_examples = 0
    samples = []

    # Hors-HSK : array de dict
    print("📖 Hors-HSK")
    horshsk_chunks = sorted(HORS_HSK_DIR.glob("chunk-*.json"))
    for c in horshsk_chunks:
        data = json.loads(c.read_text())
        n_e, n_x = process_entries(data, args.target, samples)
        if not args.dry_run and (n_e > 0 or n_x > 0):
            c.write_text(json.dumps(data, ensure_ascii=False))
        if n_e > 0 or n_x > 0:
            print(f"   {c.name}: {n_e} entrée(s), {n_x} exemple(s)")
        total_entries += n_e
        total_examples += n_x

    # HSK : map { id: entry }
    for lvl in ["hsk1", "hsk2", "hsk3", "hsk4", "hsk5", "hsk6", "hsk7"]:
        lvl_dir = HSK_DIR / lvl
        if not lvl_dir.exists():
            continue
        print(f"📚 {lvl}")
        for c in sorted(lvl_dir.glob("chunk-*.json")):
            data = json.loads(c.read_text())
            entries = list(data.values())
            n_e, n_x = process_entries(entries, args.target, samples)
            if not args.dry_run and (n_e > 0 or n_x > 0):
                # data est déjà modifié in-place puisque c'est les mêmes refs
                c.write_text(json.dumps(data, ensure_ascii=False))
            if n_e > 0 or n_x > 0:
                print(f"   {c.name}: {n_e} entrée(s), {n_x} exemple(s)")
            total_entries += n_e
            total_examples += n_x

    print()
    print("Résumé :")
    print(f"  Entrées corrigées : {total_entries}")
    print(f"  Exemples corrigés : {total_examples}")
    print()
    if samples:
        print("Exemples de modifications :")
        for s in samples:
            label = s["hanzi"] if s["kind"] == "entry" else s["chinese"]
            print(f"  [{s['kind']}] {label}")
            print(f"    avant : {s['old']}")
            print(f"    après : {s['new']}")


if __name__ == "__main__":
    main()
